### PR TITLE
Adds reason phrase support

### PR DESCRIPTION
**Pull request by sbleon**
Reason phrases are sometimes used to report additional information about why a request failed, so they're very useful in troubleshooting failed request. Building on @davidbiehl's work mentioned in https://github.com/lostisland/faraday/issues/418, this adds reason phrase support to all adapters (except Rack).

@mislav, you said in #418 that you'd be happy to see a pull request about this, so here you go!

### CODE REVIEW

**Comment by sbleon** [comment]
Looks like I missed Typhoeus due to testing on Ruby 2.2. One sec...

**Comment by sbleon** [comment]
This is good to go!

**Comment by davidbiehl** [comment]
@sbleon that's awesome! Thanks for your help with getting this implemented with the rest of the adapters :)

**Comment by sbleon** [comment]
@mislav sorry to bug you about this, but I would love to get back on a proper release of this gem. Can you take a look at this PR, please, or is this another maintainer I should bother?

File: test/adapters/rack_test.rb [hunk 1]
```diff
@@ -14,6 +14,9 @@ def adapter_options
     include Integration::Common
     include Integration::NonParallel
 
+    # Rack::MockResponse doesn't provide any way to access the reason phrase
+    undef :test_GET_reason_phrase
```
Comment (mislav) [inline test/adapters/rack_test.rb]: Could we add a test that the reason phrase is nil for this adapter?

File: lib/faraday/adapter.rb [hunk 2]
```diff
@@ -34,9 +34,10 @@ def call(env)
       env.clear_body if env.needs_body?
     end
 
-    def save_response(env, status, body, headers = nil)
+    def save_response(env, status, body, headers = nil, reason_phrase = "")
       env.status = status
       env.body = body
+      env.reason_phrase = reason_phrase.to_s.strip
```
Comment (mislav) [inline lib/faraday/adapter.rb]: Could the reason phrase be nil if it's missing, instead of an empty string? That way users can test the presence of the phrase more easily.

File: lib/faraday/options.rb [hunk 3]
```diff
@@ -252,7 +252,8 @@ def new_builder(block)
   end
 
   class Env < Options.new(:method, :body, :url, :request, :request_headers,
-    :ssl, :parallel_manager, :params, :response, :response_headers, :status)
+    :ssl, :parallel_manager, :params, :response, :response_headers, :status,
+    :reason_phrase)
```
Comment (mislav) [inline lib/faraday/options.rb]: I'm not ecstatic about the word `reason_phrase`. Could we use `status_text` instead?

**Comment by mislav** [comment]
Sorry for the late review; and thanks for the contribution!

File: test/adapters/rack_test.rb (hunk 1 above)
Comment (sbleon) [inline test/adapters/rack_test.rb]: Done in 0818a1e.

File: lib/faraday/adapter.rb (hunk 2 above)
Comment (sbleon) [inline lib/faraday/adapter.rb]: Good idea! Done in 0818a1e.

File: lib/faraday/options.rb (hunk 3 above)
Comment (sbleon) [inline lib/faraday/options.rb]: @mislav, I think that `status_text` would be a little more intuitive, as "Status Text" seems to be a common colloquial term. (E.g. https://developer.mozilla.org/en-US/docs/Web/HTTP/Response_codes).

However, "Reason Phrase" is the proper technical term per http://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html, so I'd prefer to keep the name `reason_phrase`, if that's alright with you.

What do you think?

**Comment by sbleon** [comment]
@mislav this PR should be good to, pending your approval of the term "Reason Phrase" per https://github.com/lostisland/faraday/pull/547#discussion_r48784710. Let me know if you really want me to change it to `status_text`. Otherwise, I'd love to get this merged. Thanks!

**Comment by sbleon** [comment]
Hey, @mislav ! I'm sure you're busy, but please respond to my comment above about "Reason Phrase" vs "Status Text". IMHO, this PR is good to merge ;-)

**Comment by sbleon** [comment]
Ping. I'm hoping we can get this merged. It should be good to go.

I really need to get this merged so I can send a PR to [VCR](https://github.com/vcr/vcr) that will make it support reason phrases when hooked into Faraday. Thanks!

**Comment by sbleon** [comment]
@jingweno, @technoweenie, @pengwynn, @sferik, I'm reaching out to you as members of the lostisland org. I'd really like to get this PR merged, but it seems like @mislav is really busy. I'm hoping that one of you is a committer and can help me get this merged.

Sorry to rope so many of you into this. I'm just trying to reduce the number of forks in my Gemfile!

**Comment by sbleon** [comment]
@mislav this PR should be good to go, pending your approval of the term "Reason Phrase" per https://github.com/lostisland/faraday/pull/547#discussion_r48784710. Let me know if you really want me to change it to `status_text` and I'll defer to you. Otherwise, I'd love to get this merged. Thanks!

**Comment by davidbiehl** [comment]
Another thought: get the best of both worlds by aliasing reason_phrase with
status_text. That way you have what's correct per the RFC and also the
colloquial term.

**Comment by sbleon** [comment]
Yeah, I’ve considered that as well. I prefer just `reason_phrase` because
it will help to educate developers about the proper terminology, but Ruby
does love its method aliases, so I’m open to having both.

**Comment by sbleon** [comment]
I would really, really, love to see this merged and released. @sferik @technoweenie @mislav can we move this forward, please?

**Comment by sbleon** [comment]
@iMacTia can you merge this for me? I addressed all of @mislav's code review feedback a while ago.

**Comment by iMacTia** [comment]
This was reviewed an is passing all tests. Moreover, it was asked many times in other issues. Finally, is backward-compatible as it is simply adding more info to ENV.
Good job!

Merging :)
